### PR TITLE
Animations

### PR DESCRIPTION
**Pull request by joekarasek**
## Summary

Proof of concept approach to reusable/recomposable/authorable animations for the Bolt Design System.

## Details

This PR creates a new package, `bolt/animations`. This PR takes a design system approach to animations by setting out animation design tokens/settings, tools (mixins), and utilities (animation triggers and scroll triggers). Specific animation content in this PR (timing functions, durations, delays, keyframes, etc.) are for demonstration purposes only and are not yet approved Bolt design features. Existing bespoke animation is not impacted by this PR.

## How to test

See http://localhost:3001/pattern-lab/?p=viewall-visual-styles-animations

### CODE REVIEW

File: packages/animations/utilities/animation-utils.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    //
+    // el.dispatchEvent(event);
+  }
+}
+
+/**
+ * @param {HTMLElement} el - The element
+ * @return {boolean}
+ */
+export function isElementInViewport(el) {
```
Comment (sghoweri) [inline packages/animations/utilities/animation-utils.js]: @joekarasek I’d strongly recommend looking into using IntersectionObserver for this + your `watchForOnViewAnims` function below. 
https://www.smashingmagazine.com/2018/01/deferring-lazy-loading-intersection-observer-api

There’s also a Polyfill so newer browsers can get the best native performance and older ones get the same overall performance as DIYing (or better) 😉

**Comment by mikemai2awesome** [comment]
Hi, @joekarasek, nice start one this!

I have read through the docs and have the following comments:

1. We now have coding standards documented, please read through all of it and try to follow them: https://github.com/bolt-design-system/bolt/wiki/Code-Standards-and-Conventions. For example, scss variable, function, or mixin names should not have `--`
2. Can we have the mixins just be `bolt-animation()` and `bolt-transition()`? For example, to use it, it would simply be `@include bolt-animation(fade-in)`.
3. I am really confused about the `$base` and `$active`, not sure if this is a documentation issue or naming issue.
4. Duration, timing-function, delay, direction, etc. These params that go into the mixins should all be kept as a map, similar to how we handle breakpoint scale, spacing scale, etc. Each scale should have its own function ready for use, for example, `bolt-duration(fast)`.
    - Related note: functions are kept in settings and mixins are kept in tools. But for work in progress, you can keep everything in one file as you are doing for now.
5. For delay and duration, let's use `short`, `medium`, `long` as the language for consistency.
6. Instead of using data attributes, we should be thinking about web components. Also let's be clear and not abbreviate, and follow our conventions for boolean values. `animation-triggered` should be a boolean prop and not require the yes/no values.
7. Each animation/transition should have positions attached (top, right, bottom, left), for example, `bolt-animation($name: fade-in, $position: left)` (fade in from the left side) or `bolt-animation($name: fade-out, $position: top)` (fade out to the top).
8. The final ideal mixin should look something like this:
```scss
// This element is animating.
.element {
  @include bolt-animation($name: fade-in, $duration: long, $delay: short, $timing-function: ease-in-out, $play-state: running);
}

// This element animates on hover.
.element {
  @include bolt-transition($name: fade-out, $duration: long, $delay: short, $timing-function: ease-in-out, $interaction: hover);
}
```

cc @sghoweri @danielamorse

**Comment by margoromo** [comment]
@mikemai2awesome can you take a peek here?

**Comment by mikemai2awesome** [comment]
@margoromo I did. Look above for all of my notes.

**Comment by margoromo** [comment]
Created http://vjira2:8080/browse/BDS-1716 and http://vjira2:8080/browse/BDS-1717 for continued review by @mikemai2awesome and @sghoweri

**Pull request merged**